relay-event-schema/src/protocol/nel.rs

```diff
@@ -0,0 +1,251 @@
+    #[error("incoming json is unparsable")]
+    InvalidJson(#[from] serde_json::Error),
+}
+
+/// Generated network error report (NEL).
+#[derive(Clone, Debug, Default, PartialEq, Empty, FromValue, IntoValue, ProcessValue)]
+#[cfg_attr(feature = "jsonschema", derive(JsonSchema))]
+pub struct Body {
+    /// The elapsed number of milliseconds between the start of the resource
+    /// fetch and when it was completed or aborted by the user agent.
+    pub elapsed_time: Annotated<u64>,
+    /// HTTP method.
+    pub method: Annotated<String>,
+    /// If request failed, the phase of its network error. If request succeeded, "application".
+    pub phase: Annotated<String>,
+    /// The HTTP protocol and version.
+    pub protocol: Annotated<String>,
+    /// Request's referrer, as determined by the referrer policy associated with its client.
+    pub referrer: Annotated<String>,
+    /// The sampling rate.
+    pub sampling_fraction: Annotated<f64>,
+    /// The IP address of the server where the site is hosted.
+    pub server_ip: Annotated<String>,
+    /// HTTP status code.
+    pub status_code: Annotated<i64>,
+    /// If request failed, the type of its network error. If request succeeded, "ok".
+    #[metastructure(field = "type")]
+    pub ty: Annotated<String>,
+}
+
+/// Models the content of a NEL report.
+///
+/// NOTE: This is the structure used inside the Event (serialization is based on Annotated
+/// infrastructure). We also use a version of this structure to deserialize from raw JSON
+/// via serde.
+///
+/// See <https://w3c.github.io/network-error-logging/>
+#[derive(Clone, Debug, Default, PartialEq, Empty, FromValue, IntoValue, ProcessValue)]
+#[cfg_attr(feature = "jsonschema", derive(JsonSchema))]
+pub struct NetworkReport {
+    /// The age of the report since it got collected and before it got sent.
+    pub age: Annotated<i64>,
+    /// The type of the report.
+    #[metastructure(field = "type")]
+    pub ty: Annotated<String>,
+    /// The URL of the document in which the error occurred.
+    #[metastructure(pii = "true")]
+    pub url: Annotated<String>,
+    /// The User-Agent HTTP header.
+    pub user_agent: Annotated<String>,
+    /// The body of the NEL report.
+    pub body: Annotated<Body>,
+}
+
+impl NetworkReport {
+    fn get_request(&self) -> Request {
+        let headers = match self.user_agent.value() {
+            Some(ref user_agent) if !user_agent.is_empty() => {
+                Annotated::new(Headers(PairList(vec![Annotated::new((
+                    Annotated::new(HeaderName::new("User-Agent")),
+                    Annotated::new(HeaderValue::new(user_agent)),
+                ))])))
+            }
+            Some(_) | None => Annotated::empty(),
+        };
+        Request {
+            url: self.url.clone(),
+            headers,
+            ..Request::default()
+        }
+    }
+    pub fn apply_to_event(data: &[u8], event: &mut Event) -> Result<(), ReportError> {
```

Comment:
Modelling this after CSP unfortunately means quite a bit of normalization logic is now placed in the `event-schema` crate. Could we move this to `event-normalization`, instead?